### PR TITLE
tests: gpio: add arduino gpio test

### PR DESCRIPTION
**Pull request by hakehuang**
add a arduino gpio test for all arduino supported boards user need loopback D0 - D1.

this test application is to use the unified test interface (ARDUINO header) such as no more board overlay is required.
boards with arduino interface can short D0 and D1 plus the arduino_gpio fixture in board map file. In this way, test application will be easy to expand to other paltforms.

I have to exclude below 3 boards, because they actually use mikrobus need a convert board to arduino.

related to : https://github.com/zephyrproject-rtos/zephyr/issues/68400

### CODE REVIEW

**Comment by hakehuang** [comment]
test pass on frdm_k64f

```
** Booting Zephyr OS build v3.6.0-23-g5644c38b876e ***
Running TESTSUITE gpio_arduino
===================================================================
START - test_arduino_gpio_transfer
tx start
rx start
tx data
rx data
tx data
rx data
tx data
rx data
tx data
rx data
tx data
rx data
tx data
rx data
tx data
rx data
tx parity
stop bit
rx = 55
tx = 55
 PASS - test_arduino_gpio_transfer in 0.034 seconds
===================================================================
TESTSUITE gpio_arduino succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [gpio_arduino]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.034 seconds
 - PASS - [gpio_arduino.test_arduino_gpio_transfer] duration = 0.034 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

**Comment by hakehuang** [comment]
@henrikbrixandersen @mnkp please kindly reviw this PR. thanks.

**Review by henrikbrixandersen — DISMISSED** [review]
I am not sure I understand the intention. Is this to verify (parts of) an Arduino Uno R3 connector DTS definition?

**Comment by hakehuang** [comment]
> I am not sure I understand the intention. Is this to verify (parts of) an Arduino Uno R3 connector DTS definition?

@henrikbrixandersen this is part of jobs for #https://github.com/zephyrproject-rtos/zephyr/issues/68400 , we are propose to test peripheral in a unified way, for gpio, we will using common header (arduino) here to test.

File: tests/drivers/gpio/gpio_arduino/testcase.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+common:
+  tags:
+    - drivers
+    - gpio
+  depends_on:
+    - gpio
+    - arduino_gpio
```
Comment (PerMac) [inline tests/drivers/gpio/gpio_arduino/testcase.yaml:7]: are both features required? I see boards which supports only arduino_gpio but not gpio (not sure it it is intended) e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/nxp/mimxrt1010_evk/mimxrt1010_evk.yaml

Comment (hakehuang) [reply tests/drivers/gpio/gpio_arduino/testcase.yaml:7]: this is the board yaml issue, gpio is a must. let me create another PR to fix

Comment (hakehuang) [reply tests/drivers/gpio/gpio_arduino/testcase.yaml:7]: pr created #70202 70202

**Comment by hakehuang** [comment]
@anangl , please help to review thanks

**Comment by hakehuang** [comment]
@PerMac , @anangl  can you help to review? the issue you point out is fixed in https://github.com/zephyrproject-rtos/zephyr/pull/70202 Thanks

**Review by dleach02 — DISMISSED** [review]

File: tests/drivers/gpio/gpio_arduino/testcase.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  harness_config:
+    fixture: arduino_gpio
+
+tests:
+  drivers.gpio.arduino:
+    filter: not dt_compat_enabled("arduino-header-r3")
+    integration_platforms:
+      - frdm_k64f
+    platform_exclude:
+      # below boards are mifro bus not arduino
```
Comment (nashif) [inline tests/drivers/gpio/gpio_arduino/testcase.yaml:18]: what is mifro?

File: tests/drivers/gpio/gpio_arduino/README.txt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+Title: test_gpio_arduino
```
Comment (nashif) [inline tests/drivers/gpio/gpio_arduino/README.txt:1]: that readme is very barebones and has some custom layout , can you please follow format of existing READMEs coming with tests...

File: tests/drivers/gpio/gpio_arduino/testcase.yaml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  depends_on:
+    - gpio
+    - arduino_gpio
+  harness: ztest
+  harness_config:
+    fixture: arduino_gpio
+
+tests:
+  drivers.gpio.arduino:
+    filter: not dt_compat_enabled("arduino-header-r3")
```
Comment (nashif) [inline tests/drivers/gpio/gpio_arduino/testcase.yaml:14]: do not understand this filter, you want to include everything but those with `arduino-header-r3` or the opposite, given that the test is about arduino

**Review by nashif — CHANGES_REQUESTED** [review]

File: tests/drivers/gpio/gpio_arduino/testcase.yaml (hunk 2 above)
Comment (hakehuang) [reply tests/drivers/gpio/gpio_arduino/testcase.yaml:18]: sorry typo mikrobus, let me fix

File: tests/drivers/gpio/gpio_arduino/testcase.yaml (hunk 4 above)
Comment (hakehuang) [reply tests/drivers/gpio/gpio_arduino/testcase.yaml]: sorry this is a debug code. I need change back

File: tests/drivers/gpio/gpio_arduino/README.txt (hunk 3 above)
Comment (hakehuang) [reply tests/drivers/gpio/gpio_arduino/README.txt]: Thanks for checking this, I am too lazy here. updated

**Comment by hakehuang** [comment]
@nashif  update the document according to other tests. Please check thanks in advance. also optimize the testcase.yaml for more compatible with all platforms.
@henrikbrixandersen , @dleach02  Thanks for reviewing, I update according to anash comments, please revist.

**Comment by hakehuang** [comment]
@nashif please help to review, thanks a lot.

**Review by gmarull — CHANGES_REQUESTED** [review]
what is the exact value of this test? there's already a suite covering most/all of the GPIO API

**Comment by PerMac** [comment]
If I see correctly, one needs to shorten 2 pins on arduino shield. However, I can do this already at the dk without arduino. Why would anyone bother to add arduino into the picture?

**Comment by hakehuang** [comment]
> what is the exact value of this test? there's already a suite covering most/all of the GPIO API

@gmarull , this is an approach to unified the test, in this case any arduino supported board will not need to add any dts to support gpio test cases, it will be supported if an arduino interface is defined. and this can applies to i2c/spi/uart as well

**Comment by gmarull** [comment]
> > what is the exact value of this test? there's already a suite covering most/all of the GPIO API
> 
> @gmarull , this is an approach to unified the test, in this case any arduino supported board will not need to add any dts to support gpio test cases, it will be supported if an arduino interface is defined. and this can applies to i2c/spi/uart as well

I'm pretty sure there's a better way to achieve this other than adding a new full-blown test. My -1 stands.

**Comment by PerMac** [comment]
In my understanding we want a test itself to be hw independent and hw dependencies included on an additional configuration variants layer. We shouldn't duplicate tests for the same functionality but rather to add more configurations to the existing tests (not replacing).

**Pull request closed without merging**